src/bin/pg_autoctl/service_postgres_ctl.c

```diff
@@ -0,0 +1,511 @@
+/*
+ * src/bin/pg_autoctl/postgres_service_ctl.c
+ *   Utilities to start/stop the pg_autoctl service.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+
+#include <inttypes.h>
+#include <limits.h>
+#include <sys/select.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "cli_common.h"
+#include "cli_root.h"
+#include "defaults.h"
+#include "log.h"
+#include "monitor.h"
+#include "monitor_config.h"
+#include "pgsetup.h"
+#include "primary_standby.h"
+#include "service.h"
+#include "service_postgres.h"
+#include "service_postgres_ctl.h"
+#include "signals.h"
+#include "supervisor.h"
+#include "state.h"
+#include "string_utils.h"
+
+#include "runprogram.h"
+
+static bool shutdownSequenceInProgress = false;
+
+static bool ensure_postgres_status(LocalPostgresServer *postgres,
+								   Service *service);
+
+static bool ensure_postgres_status_stopped(LocalPostgresServer *postgres,
+										   Service *service);
+
+static bool ensure_postgres_status_running(LocalPostgresServer *postgres,
+										   Service *service);
+
+
+/*
+ * service_postgres_ctl_start starts a subprocess that implements the postgres
+ * service depending on the current assigned and goal state of the keeper.
+ */
+bool
+service_postgres_ctl_start(void *context, pid_t *pid)
+{
+	pid_t fpid;
+
+	/* Flush stdio channels just before fork, to avoid double-output problems */
+	fflush(stdout);
+	fflush(stderr);
+
+	/* time to create the node_active sub-process */
+	fpid = fork();
```

Comment:
I think I see what you mean, and I'm not sure it would be of much use. Will have to look at the source again with the other PRs to see if we would then be able to have the same function everywhere just calling the new API, but I think we have different local variables specific to each service that we want to use in each place anyway.